### PR TITLE
Use honggfuzz to fuzz firmware rather than AFL

### PR DESCRIPTION
**Pull request by hughsie**
This has better multi-core performance and can run in persistent mode -- which
allows us to construct a test harness of all the parsers (which takes time) and
then just reuse the process for lots of different data.

Type of pull request:
- [ ] New plugin (Please include [new plugin checklist](https://github.com/fwupd/fwupd/wiki/New-plugin-checklist))
- [ ] Code fix
- [X] Feature
- [ ] Documentation

### CODE REVIEW

File: src/fu-engine.h [hunk 1]
```diff
@@ -26,13 +26,18 @@ G_DECLARE_FINAL_TYPE (FuEngine, fu_engine, FU, ENGINE, GObject)
  * FuEngineLoadFlags:
  * @FU_ENGINE_LOAD_FLAG_NONE:		No flags set
  * @FU_ENGINE_LOAD_FLAG_READONLY_FS:	Ignore readonly filesystem errors
+ * @FU_ENGINE_LOAD_FLAG_NO_ENUMERATE:	Do not enumerate devices
+ * @FU_ENGINE_LOAD_FLAG_NO_REMOTES:	Do not enumerate remotes
+ * @FU_ENGINE_LOAD_FLAG_NO_HWINFO:	Do not load details about the hardware
```
Comment (superm1) [inline src/fu-engine.h:31]: These feel like they should also be fwupdtool arguments.

Comment (hughsie) [reply src/fu-engine.h:31]: I fixed this in a different way in https://github.com/fwupd/fwupd/pull/2668

**Comment by hughsie** [comment]
@superm1 I'm happy with this now, re-review welcome, thanks.

File: meson.build [hunk 2]
```diff
@@ -113,6 +112,14 @@ warning_flags = [
 cc = meson.get_compiler('c')
 add_project_arguments(cc.get_supported_arguments(warning_flags), language : 'c')
 
+# use honggfuzz for parallel persistent fuzzing
+if cc.has_function('HF_ITER')
+  honggfuzz = find_program('honggfuzz')
+  conf.set('HAVE_HF_ITER', '1')
+else
+  honggfuzz = find_program('honggfuzz', required : false)
```
Comment (superm1) [inline meson.build:120]: I don't know if this works, but can you have:
`honggfuzz = find_program('honggfuzz', required : cc.has_function('HF_ITER')`

Comment (hughsie) [reply meson.build:120]: I still need the HAVE_HF_ITER define tho

Comment (superm1) [reply meson.build:120]: Like this maybe: 
```
honggfuzz = find_program('honggfuzz', required : cc.has_function('HF_ITER')
if honggfuzz.found()
   conf.set('HAVE_HF_ITER', '1')
endif
```

Comment (hughsie) [reply meson.build]: Works for me! Fixed.

File: meson.build [hunk 3]
```diff
@@ -113,6 +112,12 @@ warning_flags = [
 cc = meson.get_compiler('c')
 add_project_arguments(cc.get_supported_arguments(warning_flags), language : 'c')
 
+# use honggfuzz for parallel persistent fuzzing
+honggfuzz = find_program('honggfuzz', required: cc.has_function('HF_ITER'))
+if honggfuzz.found()
+  conf.set('HAVE_HF_ITER', '1')
+endif
+
```
Comment (superm1) [inline meson.build:120]: you know I just realized something about doing it this way though.  If `honggfuzz` is _installed_ that will turn on fuzzing everywhere.  is that expected?  Or does this need an extra meson option?

Comment (hughsie) [reply meson.build:120]: >  is that expected 

Yes, you can use honggfuzz without compiling with the custom clang compiler, but you can't use the persistent mode.

Comment (superm1) [reply meson.build:120]: OK.

**Review by superm1 — APPROVED** [review]

**Pull request merged**